### Machine Learning Foundation Specialization
### University of Washington - Seattle

#### Week 3 Quiz#2 Analyzing Product Sentiment

In [3]:
## Created by Quincy Gu
## Created on 03/25/2020 31:39
## Gu.Qiangqiang@mayo.edu
## Mayo Clinic College of Medicine and Sciences

#### Instruction

##### In the Jupyter notebook above, we used the word counts for all words in the reviews to train the sentiment classifier model. Now, we are going to follow a similar path, but only use this subset of the words:

In [48]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

##### Often, ML practitioners will throw out words they consider “unimportant” before training their model. This procedure can often be helpful in terms of accuracy. Here, we are going to throw out all words except for the very few above. Using so few words in our model will hurt our accuracy, but help us interpret what our classifier is doing.

In [260]:
import turicreate

In [261]:
## Load amazon_baby Data
products = turicreate.SFrame('amazon_baby.sframe')

## View head of the loaded data frame
products

Columns:
	name	str
	review	str
	rating	float

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[183531 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

##### 1. **Use .apply() to build a new feature with the counts for each of the selected_words:** In the notebook above, we created a column ‘word_count’ with the word counts for each review. Our first task is to create a new column in the products SFrame with the counts for each selected_word above, and, in the process, we will see how the method .apply() can be used to create new columns in our data (our features) and how to use a Python function, which is an extremely useful concept to grasp!

##### Our first goal is to create a column products[‘awesome’] where each row contains the number of times the word ‘awesome’ showed up in the review for the corresponding product, and 0 if the review didn’t show up. One way to do this is to look at the each row ‘word_count’ column and follow this logic:

##### * If ‘awesome’ shows up in the word counts for a particular product (row of the products SFrame), then we know how often ‘awesome’ appeared in the review.
##### * If ‘awesome’ doesn’t appear in the word counts, then it didn’t appear in the review, and we should set the count for ‘awesome’ to 0 in this review.

In [262]:
products['word_counts'] = turicreate.text_analytics.count_words(products['review'])
products

Columns:
	name	str
	review	str
	rating	float
	word_counts	dict

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|          word_counts          |
+-------------------------------+
| {'handles': 1.0, 'strippin... |
| {'recommend': 1.0, 'moist'... |
| {'quilt': 1.0, 'comfortabl... |
| {'tool': 1.0, 'clever': 1.... |
| {'rock': 1.0, 'many': 1.0,... |
| {'thumb': 1.0, 'or': 1.0, ... |
| {'for': 1.0, 'barnes': 1.0... |
| {'right': 1.0, 'because': ... |
| {'like': 1.0, 'and': 1.0, ... |
| {'in': 1.0, 'pages': 1.0, ... |
+-------------------------------+
[183531 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [263]:
for i in selected_words:
    products[i] = products['word_counts'].apply(lambda counts:counts.get(i,0))
products

Columns:
	name	str
	review	str
	rating	float
	word_counts	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|          word_counts          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'handles': 1.0, 'strippin... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'recommend': 1.0, 'moist'... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'quilt': 1.0, 'comfortabl... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'tool': 1.0, 'clever': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
| {'rock': 1.0, 'many': 1.0,... |   0.0   |  1.0  |    0.0    |   0.0   | 1.0  |
| {'thumb': 1.0, 'or': 1.0, ... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'for': 1.0, 'barnes': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'right': 1.0, 'because': ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'like': 1.0, 'and': 1.0, ... |   0.0   |  0.0  |    1.0    |   0.0   | 0.0  |
| {'in': 1.0, 'pages': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+
| horrible | bad | terrible | awful | wow | hate |
+----------+-----+----------+-------+-----+------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
|    0     | 0.0 |    0     |   0   |  0  |  0   |
+----------+-----+----------+-------+-----+------+
[183531 rows x 15 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

##### We could use a for loop to iterate this logic for each row of the products SFrame, but this approach would be really slow, because the SFrame is not optimized for this being accessed with a for loop. Instead, we will use the .apply() method to iterate the the logic above for each row of the products[‘word_count’] column (which, since it’s a single column, has type SArray). Read about using the .apply() method on an SArray here.

##### We are now ready to create our new columns:
#####    * First, you will use a Python function to define the logic above. You will write a function called awesome_count which takes in the word counts and returns the number of times ‘awesome’ appears in the reviews.

##### A few tips:

##### i. Each entry of the ‘word_count’ column is of Python type dictionary.

##### ii. If you have a dictionary called dict, you can access a field in the dictionary using:

In [264]:
# dict['awesome']

##### but only if ‘awesome’ is one of the fields in the dictionary, otherwise you will get a nasty error.

##### iii. In Python, to test if a dictionary has a particular field, you can simply write:

In [265]:
# if 'awesome' in dict

##### In our case, if this condition doesn’t hold, the count of ‘awesome’ should be 0.

##### Using these tips, you can now write the awesome_count function.
#####    * Next, you will use .apply() to iterate awesome_count for each row of products[‘word_count’] and create a new column called ‘awesome’ with the resulting counts. Here is what that looks like:

In [266]:
# products['awesome'] = products['word_count'].apply(awesome_count)

##### And you are done! Check the products SFrame and you should see the new column you just create.

##### * Repeat this process for the other 11 words in selected_words. (Here, we described a simple procedure to obtain the counts for each selected_word. There are other more efficient ways of doing this, and we encourage you to explore this further.)

##### * Using the .sum() method on each of the new columns you created, answer the following questions: Out of the selected_words, which one is most used in the dataset? Which one is least used? **Save these results to answer the quiz at the end.**

In [267]:
sum_selected_counts = {}
count_sums = list()
for i in range(len(selected_words)):
    count_sums.append(products[selected_words[i]].sum())
    print('The sum of total counts of word', selected_words[i], products[selected_words[i]].sum())

The sum of total counts of word awesome 4075.0
The sum of total counts of word great 59536.0
The sum of total counts of word fantastic 1765.0
The sum of total counts of word amazing 2726.0
The sum of total counts of word love 43867.0
The sum of total counts of word horrible 1245
The sum of total counts of word bad 4950.0
The sum of total counts of word terrible 1282
The sum of total counts of word awful 753
The sum of total counts of word wow 461
The sum of total counts of word hate 1285


In [268]:
sum_selected_counts = list()
for i in range(len(selected_words)):
    sum_selected_counts.append(count_sums[i])

In [269]:
most_used_selected_word = selected_words[sum_selected_counts.index(max(sum_selected_counts))]
least_used_selected_word = selected_words[sum_selected_counts.index(min(sum_selected_counts))]

print('Out of the 11 words in selected_words, the most used in the reviews in the dataset is:',most_used_selected_word,
      '\n')
print('Out of the 11 words in selected_words, the least used in the reviews in the dataset is:',least_used_selected_word,
      '\n')

Out of the 11 words in selected_words, the most used in the reviews in the dataset is: great 

Out of the 11 words in selected_words, the least used in the reviews in the dataset is: wow 



##### 2. **Create a new sentiment analysis model using only the selected_words as features:** In the Jupyter Notebook above, we used word counts for all words as features for our sentiment classifier. Now, you are just going to use the selected_words:

##### * Use the same train/test split as in the Jupyter Notebook from lecture:

In [270]:
products = products[products['rating'] != 3]
products['sentiment'] = products['rating'] >= 4

In [271]:
train_data,test_data = products.random_split(.8, seed=0)

##### Train a logistic regression classifier (use turicreate.logistic_classifier.create) using just the selected_words. Hint: you can use this parameter in the .create() call to specify the features used to be exactly the new columns you just created:

In [301]:
selected_words_model = turicreate.logistic_classifier.create(train_data,
                                                                      target = 'sentiment',
                                                                      features = selected_words,
                                                                      validation_set = test_data)

sentiment_model = turicreate.logistic_classifier.create(train_data,
                                                       target = 'sentiment',
                                                       features = ['word_counts'],
                                                       validation_set = test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.090654     | 0.847401          | 0.845874            |

| 2         | 3        | 0.147877     | 0.847514          | 0.846085            |

| 3         | 4        | 0.199700     | 0.847626          | 0.846115            |

| 4         | 5        | 0.253181     | 0.847708          | 0.846385            |

| 5         | 6        | 0.304002     | 0.847708          | 0.846385            |

| 6         | 7        | 0.357485     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 0.485293     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 1.253344     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 1.627420     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 1.860023     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 2.206574     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 3.521026     | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

##### Call your new model: selected_words_model.

##### You will now examine the weights the learned classifier assigned to each of the 11 words in selected_words and gain intuition as to what the ML algorithm did for your data using these features. In Turi Create, a learned model, such as the selected_words_model, has a field 'coefficients', which lets you look at the learned coefficients. You can access it by using:

In [302]:
selected_words_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 12

Data:
+-------------+-------+-------+---------------------+----------------------+
|     name    | index | class |        value        |        stderr        |
+-------------+-------+-------+---------------------+----------------------+
| (intercept) |  None |   1   |  1.3365913848877558 | 0.008929969787656753 |
|   awesome   |  None |   1   |  1.133534666034145  | 0.08399643983187526  |
|    great    |  None |   1   |  0.8630655001196618 | 0.018955052444377323 |
|  fantastic  |  None |   1   |  0.8858047568814295 | 0.11167591293399656  |
|   amazing   |  None |   1   |  1.1000933113660283 | 0.09954776260465983  |
|     love    |  None |   1   |  1.3592688669225153 | 0.028068300152099435 |
|   horrible  |  None |   1   |  -2.251335236759093 | 0.08020249388788442  |
|     bad     |  None |   1   | -0.9914778800650565 | 0.03848428664699063  |
|   terrible  |  None |   1   |  -2.223661436085127 | 0.07731736203785755  |
|    awful    |  None |   1   | -2.0529082040313513 | 0.10099735435259259  |
+-------------+-------+-------+---------------------+----------------------+
[12 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [303]:
selected_words_model.coefficients.sort('value', ascending=False).print_rows(num_rows=12)

+-------------+-------+-------+-----------------------+----------------------+
|     name    | index | class |         value         |        stderr        |
+-------------+-------+-------+-----------------------+----------------------+
|     love    |  None |   1   |   1.3592688669225153  | 0.028068300152099435 |
| (intercept) |  None |   1   |   1.3365913848877558  | 0.008929969787656753 |
|   awesome   |  None |   1   |   1.133534666034145   | 0.08399643983187526  |
|   amazing   |  None |   1   |   1.1000933113660283  | 0.09954776260465983  |
|  fantastic  |  None |   1   |   0.8858047568814295  | 0.11167591293399656  |
|    great    |  None |   1   |   0.8630655001196618  | 0.018955052444377323 |
|     wow     |  None |   1   | -0.009538236067678897 |  0.1604641122471166  |
|     bad     |  None |   1   |  -0.9914778800650565  | 0.03848428664699063  |
|     hate    |  None |   1   |  -1.3484407222463124  | 0.07715698604297333  |
|    awful    |  None |   1   |  -2.0529082040313513

##### The result has a column called ‘value’, which contains the weight learned for each feature.

##### Using this approach, sort the learned coefficients according to the ‘value’ column using .sort(). Out of the 11 words in selected_words, which one got the most positive weight? Which one got the most negative weight? Do these values make sense for you? **Save these results to answer the quiz at the end.**

##### 3. **Comparing the accuracy of different sentiment analysis model: Using the method**

In [306]:
sentiment_model.evaluate(test_data)

{'accuracy': 0.9176975738650012,
 'auc': 0.9342357833151299,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        1        |  1397 |
 |      0       |        0        |  3931 |
 |      1       |        1        | 26632 |
 |      1       |        0        |  1344 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.951057941255245,
 'log_loss': 0.3304787187241227,
 'precision': 0.9501587641371436,
 'recall': 0.9519588218472976,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+--------------------+-------+------+
 | threshold |        fpr         |        tpr         |   p   |  n   |
 +-----------+--------------------+--------------------+-------+------+
 |    0.0

In [307]:
selected_words_model.evaluate(test_data)

{'accuracy': 0.8463848186404036,
 'auc': 0.6936022046674926,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        0        |  371  |
 |      0       |        1        |  4957 |
 |      1       |        1        | 27817 |
 |      1       |        0        |  159  |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9157860082304526,
 'log_loss': 0.3962265467087378,
 'precision': 0.8487520595594068,
 'recall': 0.9943165570488991,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+-----+-------+------+
 | threshold |        fpr         | tpr |   p   |  n   |
 +-----------+--------------------+-----+-------+------+
 |    0.0    |        1.0         | 1.0 | 27976 | 532

##### What is the accuracy of the selected_words_model on the test_data? What was the accuracy of the sentiment_model that we learned using all the word counts in the Jupyter Notebook above from the lectures? What is the accuracy majority class classifier on this task? How do you compare the different learned models with the baseline approach where we are just predicting the majority class? **Save these results to answer the quiz at the end.**

In [308]:
num_positive  = (train_data['sentiment'] == +1).sum()
num_negative = (train_data['sentiment'] == -1).sum()

accuracy_train_data_majority = num_positive/len(train_data)
print('The ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the train_data is:',
      accuracy_train_data_majority)

The ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the train_data is: 0.8413988969486241


In [309]:
test_num_positive  = (test_data['sentiment'] == +1).sum()
test_num_negative = (test_data['sentiment'] == -1).sum()

accuracy_test_data_majority = test_num_positive/len(test_data)
print('The ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the test_data is:',
      accuracy_test_data_majority)

The ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the test_data is: 0.8400192169108815


##### Hint: we discussed the majority class classifier in lecture, which simply predicts that every data point is from the most common class. This is baseline is something we definitely want to beat with models we learn from data.

##### 4. **Interpreting the difference in performance between the models:** To understand why the model with all word counts performs better than the one with only the selected_words, we will now examine the reviews for a particular product.

##### * We will investigate a product named ‘Baby Trend Diaper Champ’. (This is a trash can for soiled baby diapers, which keeps the smell contained.)

In [310]:
products[products['name'] == 'Baby Trend Diaper Champ']

Columns:
	name	str
	review	str
	rating	float
	word_counts	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int

Rows: Unknown

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|          word_counts          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'convenient': 1.0, 'can':... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'try': 1.0, 'going': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'nose': 1.0, 'for': 2.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'out': 1.0, 'pull': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'winter': 1.0, 'especiall... |   0.0   |  0.0  |    0.0    |   1.0   | 0.0  |
| {'mom': 1.0, 'huge': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'clean': 1.0, 'it': 8.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'best': 1.0, 'that': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'need': 1.0, 'garbage': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    1     | 0.0 |    0     |   0   |  1  |  0   |     1     |
|    0     | 1.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
[? rows x 16 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

##### * Just like we did for the reviews for the giraffe toy in the Jupyter Notebook in the lecture video, before we start our analysis you should select all reviews where the product name is ‘Baby Trend Diaper Champ’. Let’s call this table diaper_champ_reviews.

In [311]:
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']

##### * Again, just as in the video, use the sentiment_model to predict the sentiment of each review in diaper_champ_reviews and sort the results according to their ‘predicted_sentiment’.

In [313]:
diaper_champ_reviews['predicted_sentiment'] = sentiment_model.predict(diaper_champ_reviews,
                                                                     output_type = 'probability')
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_counts	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|          word_counts          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'convenient': 1.0, 'can':... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'try': 1.0, 'going': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'nose': 1.0, 'for': 2.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'out': 1.0, 'pull': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'winter': 1.0, 'especiall... |   0.0   |  0.0  |    0.0    |   1.0   | 0.0  |
| {'mom': 1.0, 'huge': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'clean': 1.0, 'it': 8.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'best': 1.0, 'that': 1.0,... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'need': 1.0, 'garbage': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    1     | 0.0 |    0     |   0   |  1  |  0   |     1     |
|    0     | 1.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     0     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+-----------------------+
|  predicted_sentiment  |
+-----------------------+
|   0.9950122935570309  |
|  5.9014145328826e-13  |
|   0.9999996193960345  |
|   0.9999567191544851  |
|   0.9999997013199228  |
|   0.9999920758179407  |
| 0.0017908398242892259 |
|   0.9999999999465673  |
|   0.9801174413554099  |
|   0.9999454483919017  |
+-----------------------+
[10 rows x 17 columns]

In [315]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment',ascending = False)

In [316]:
diaper_champ_reviews.head()

Columns:
	name	str
	review	str
	rating	float
	word_counts	dict
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	float
	terrible	int
	awful	int
	wow	int
	hate	int
	sentiment	int
	predicted_sentiment	float

Rows: 10

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | I read a review below that... |  4.0   |
| Baby Trend Diaper Champ | I have never written a rev... |  5.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | Baby Luke can turn a clean... |  5.0   |
| Baby Trend Diaper Champ | Diaper Champ or Diaper Gen... |  5.0   |
| Baby Trend Diaper Champ | I am one of those super-cr... |  5.0   |
| Baby Trend Diaper Champ | I LOOOVE this diaper pail!... |  5.0   |
| Baby Trend Diaper Champ | As a first time mother, I ... |  5.0   |
| Baby Trend Diaper Champ | I see that there are compl... |  5.0   |
| Baby Trend Diaper Champ | I have a 10 year old daugh... |  5.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+---------+-------+-----------+---------+------+
|          word_counts          | awesome | great | fantastic | amazing | love |
+-------------------------------+---------+-------+-----------+---------+------+
| {'key': 1.0, 'have': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'priceless': 1.0, 'parent... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'price': 1.0, 'worth': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'around': 1.0, 'bulk': 1.... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'be': 1.0, 'won': 1.0, 'o... |   0.0   |  1.0  |    0.0    |   0.0   | 0.0  |
| {'helps': 1.0, 'slower': 1... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'buy': 1.0, 'looking': 1.... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'ll': 1.0, 'baby': 1.0, '... |   0.0   |  0.0  |    0.0    |   0.0   | 1.0  |
| {'very': 1.0, 'seems': 1.0... |   0.0   |  0.0  |    0.0    |   0.0   | 0.0  |
| {'sorry': 1.0, 'be': 1.0, ... |   0.0   |  0.0  |    0.0    |   0.0   | 2.0  |
+-------------------------------+---------+-------+-----------+---------+------+
+----------+-----+----------+-------+-----+------+-----------+
| horrible | bad | terrible | awful | wow | hate | sentiment |
+----------+-----+----------+-------+-----+------+-----------+
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
|    0     | 0.0 |    0     |   0   |  0  |  0   |     1     |
+----------+-----+----------+-------+-----+------+-----------+
+---------------------+
| predicted_sentiment |
+---------------------+
|  0.9999999999895941 |
|  0.9999999999868131 |
|  0.9999999999465673 |
|  0.9999999999302822 |
|  0.9999999999174133 |
|  0.9999999998430963 |
|  0.9999999997360196 |
|  0.9999999995664315 |
|  0.9999999985015902 |
|  0.999999998056851  |
+---------------------+
[10 rows x 17 columns]

##### * What is the ‘predicted_sentiment’ for the most positive review for ‘Baby Trend Diaper Champ’ according to the sentiment_model from the Jupyter Notebook from lecture? **Save this result to answer the quiz at the end.**

##### * Now use the selected_words_model you learned using just the selected_words to predict the sentiment most positive review you found above. Hint: if you sorted the diaper_champ_reviews in descending order (from most positive to most negative), this command will be helpful to make the prediction you need:

In [317]:
selected_words_model.predict(diaper_champ_reviews[0:1], output_type='probability')

dtype: float
Rows: 1
[0.7919288370624453]

##### **Save this result to answer the quiz at the end.**

##### * Why is the predicted_sentiment for the most positive review found using the model with all word counts (sentiment_model) much more positive than the one using only the selected_words (selected_words_model)? Hint: examine the text of this review, the extracted word counts for all words, and the word counts for each of the selected_words, and you will see what each model used to make its prediction.**Save this result to answer the quiz at the end.**

<br/>

##### Quiz Problems and Answers

##### **Problem 1**. Out of the 11 words in selected_words, which one is most used in the reviews in the dataset?
##### **Answer:** great

##### **Problem 2**. Out of the 11 words in selected_words, which one is least used in the reviews in the dataset?
##### **Answer:** wow

##### **Problem 3**. Out of the 11 words in selected_words, which one got the most positive weight in the selected_words_model? (Tip: when printing the list of coefficients, make sure to use print_rows(rows=12) to print ALL coefficients.)
##### **Answer:** love

##### **Problem 4**. Out of the 11 words in selected_words, which one got the most negative weight in the selected_words_model? (Tip: when printing the list of coefficients, make sure to use print_rows(rows=12) to print ALL coefficients.)
##### **Answer:** horrible

##### **Problem 5**. Which of the following ranges contains the accuracy of the selected_words_model on the test_data?
##### **Answer:** 0.841 to 0.871

##### **Problem 6**. Which of the following ranges contains the accuracy of the sentiment_model in the IPython Notebook from lecture on the test_data?
##### **Answer:** 0.901 to 0.931

##### **Problem 7**. Which of the following ranges contains the accuracy of the majority class classifier, which simply predicts the majority class on the test_data?
##### **Answer:** 0.811 to 0.843

##### **Problem 8**. How do you compare the different learned models with the baseline approach where we are just predicting the majority class?
##### **Answer:** Predicting the simply majority class performed much better than the other two models.

##### **Problem 9**. Which of the following ranges contains the ‘predicted_sentiment’ for the most positive review for ‘Baby Trend Diaper Champ’, according to the sentiment_model from the IPython Notebook from lecture?
##### **Answer:** 0.9 to 1.0

##### **Problem 10**. Consider the most positive review for ‘Baby Trend Diaper Champ’ according to the sentiment_model from the IPython Notebook from lecture. Which of the following ranges contains the predicted_sentiment for this review, if we use the selected_words_model to analyze it?
##### **Answer:** 0.7 to 0.8

##### **Problem 11**. Why is the value of the predicted_sentiment for the most positive review found using the sentiment_model much more positive than the value predicted using the selected_words_model?
##### **Answer:** None of the selected_words appeared in the text of this review.